### ISSUE TITLE
Strange effect on display png images with transparent background.

### ISSUE THREAD

**Issue by mrhdias** [kivy/kivy#4463]
When I display the images, they have a dark border around, but the issue only happen in iphone and ipad. I created the images with GIMP.

Is there any solution to solve this problem?

Partial screenshot form ipad
![issue](https://cloud.githubusercontent.com/assets/589045/16761807/bd576a14-4819-11e6-8b24-66b126d3348e.png)

```
from kivy.app import App
from kivy.graphics import Color, Rectangle
from kivy.uix.gridlayout import GridLayout
from kivy.uix.image import Image

class ShowImage(GridLayout):
    def __init__(self, *args, **kwargs):
        super(ShowImage, self).__init__(**kwargs)

        self.size_hint = (1, 1)
        self.rows = 1
        self.cols = 3

        self.add_widget(Image(size_hint=(1/3., 1), source='img01.png'))
        self.add_widget(Image(size_hint=(1/3., 1), source='img02.png'))
        self.add_widget(Image(size_hint=(1/3., 1), source='img03.png'))

        self.rect = None
        with self.canvas.before:
            Color(1, 1, 1, 1, mode='rgba')
            self.rect = Rectangle(pos=self.pos, size=self.size)

        def update_pos(instance, pos):
            self.rect.pos = pos
        def update_size(instance, size):
            self.rect.size = size
        self.bind(pos=update_pos)
        self.bind(size=update_size)

class TestApp(App):
    def build(self):
        return ShowImage()

if __name__ == '__main__':
    TestApp().run()
```

![img01](https://cloud.githubusercontent.com/assets/589045/16762138/2052c752-481b-11e6-9900-ee26e40ee036.png)
![img02](https://cloud.githubusercontent.com/assets/589045/16762137/2052a740-481b-11e6-92e7-a67cf4f1302e.png)
![img03](https://cloud.githubusercontent.com/assets/589045/16762140/207bf82a-481b-11e6-8341-37890c9f2241.png)

**Comment by mrhdias** [comment]
To fix this issue in Android, I add the "png" recipe to the requirements in buildozer.spec. But I can not find a solution for iOS.

**Comment by akshayaurora** [comment]
can you try setting KIVY_IMAGE=sdl2,pil env var before any of your kivy imports. Make sure to do this before any kivy import.

**Comment by mrhdias** [comment]
I tried to compile the kivy-ios with 'pil' support, but failed
$ ./toolchain.py build pil pyobjus plyer openssl pycrypto kivy

```
...
Install frameworks for libjpeg
Install_frameworks libjpeg
Install sources for libjpeg
Install_sources libjpeg
Install libjpeg
Install libjpeg
# (ignored) Download openssl
# (ignored) Extract openssl
# (ignored) Build_all openssl
# (ignored) Download sdl2
# (ignored) Extract sdl2
# (ignored) Build_all sdl2
# (ignored) Download hostpython
# (ignored) Extract hostpython
# (ignored) Build_all hostpython
# (ignored) Download sdl2_image
# (ignored) Extract sdl2_image
# (ignored) Build_all sdl2_image
# (ignored) Download sdl2_mixer
# (ignored) Extract sdl2_mixer
# (ignored) Build_all sdl2_mixer
# (ignored) Download sdl2_ttf
# (ignored) Extract sdl2_ttf
# (ignored) Build_all sdl2_ttf
Download host_setuptools
Extract host_setuptools
Extract host_setuptools for i386
Error: cannot detect root directory
Unrecognized extension for /Users/prrcarvalho/Development/kivy-ios-master/.cache/host_setuptools-
Traceback (most recent call last):
  File "./toolchain.py", line 1311, in <module>
    ToolchainCL()
  File "./toolchain.py", line 1082, in __init__
    getattr(self, args.command)()
  File "./toolchain.py", line 1106, in build
    build_recipes(args.recipe, ctx)
  File "./toolchain.py", line 974, in build_recipes
    recipe.execute()
  File "/Users/prrcarvalho/Development/kivy-ios-master/toolchain.py", line 585, in execute
    self.extract()
  File "/Users/prrcarvalho/Development/kivy-ios-master/toolchain.py", line 56, in _cache_execution
    f(self, *args, **kwargs)
  File "/Users/prrcarvalho/Development/kivy-ios-master/toolchain.py", line 621, in extract
    self.extract_arch(arch.arch)
  File "/Users/prrcarvalho/Development/kivy-ios-master/toolchain.py", line 625, in extract_arch
    dest_dir = join(build_dir, self.archive_root)
  File "/Users/prrcarvalho/Development/kivy-ios-master/toolchain.py", line 577, in archive_root
    value = self.get_archive_rootdir(self.archive_fn)
  File "/Users/prrcarvalho/Development/kivy-ios-master/toolchain.py", line 458, in get_archive_rootdir
    raise Exception()
Exception
```

**Comment by akshayaurora** [comment]
PIL is not really needed unless you are loading a gif. Just setting KIVY_IMAGE=sdl2 should work too.

**Comment by mrhdias** [comment]
Before any of your kivy imports I add this two lines of code:

```
import os
os.environ['KIVY_IMAGE'] = 'sdl2'
```

and I get this error message :
Exception: Unknown <gif> type, no loader found.

**Comment by mrhdias** [comment]
I think that i found the image:
https://github.com/kivy/kivy/blob/master/kivy/data/images/image-loading.gif

**Comment by mrhdias** [comment]
I replace the image https://github.com/kivy/kivy/blob/master/kivy/data/images/image-loading.gif with a png image and set the environment variable  KIVY_IMAGE=sdl2. When I display the images, they have a dark border around. The sdl2 don't fix the issue.

**Comment by tito** [comment]
Just to let you know, i got that too. Your issue is from the created gimp image + OpenGL Blending + OpenGL texture filter.

The issue is: when you stretch an image, the alpha doesn't match the rgb on the same pixels. Blending happen, and it does not do what you which it does. Let me explain you in detail.

# Image that doesn't work

- Create a full transparent image
- Draw a circle in the middle of it in white
- Now analyse the RGBA layers:
  - RGB have white only where you draw: it has only white and black values
  - A is full opacity on the circle, and smoothing the edge, fully transparent outside the circle: it has a full range of 0-255 values.

Now stretch that.

- RGB will blend between pixels, and the "generated" texture will contain
- Alpha will blend too, but it will blend where there is "black" basically.

= You see the glitch from between pixels.

# How to fix the image

- On the RGB layer, extend the "white" outside the circle. You won't see it due to the Alpha channel

Then when you stretch the image, there wont be any blending on RGB, and so alpha blended pixels will always match white, not black.

# Conclusion

Image works correctly when used at scale 1. You need to fix your image for upper scale. :)

**Comment by princeIta** [comment]
Hi. I don't get you. I'm quite new to kivy. I'm working on an app for Android, on it it I need to display a png image with some transparency but on my Android mobile a black rectangle is displayed in the position where the image is supposed to be. This same image is used as the splash screen of the app and it works just fine but within the app it shows all black. The image also displays fine in my laptop

**Comment by tshirtman** [comment]
@princeIta that's a different issue, it's possible that your image is bigger than the maximum opengl texture on your phone, try downscaling it. (to 2048x2048 max).

**Comment by princeIta** [comment]
Thanks so much for the response @tshirtman. I Tried down scaling but it  did not work. Every other png or jpg image displays when I run my app on phone, but this particular image doesn't. It is able to display on the apps splash screen. And the image happens to be the logo I intend to use for my application. The image is a write-up in black colour and I'm trying to put it on a white background.

**Comment by tshirtman** [comment]
can you open a new issue about that, and put the image (or another showing the same issue) in it?

**Comment by KingValmond** [comment]
Hi,

I got this problem (I guess) too.
What I figured out, and what fixed it, is that when blending alpha-pixels, the alpha-pixels color isn't multiplyed with the alpha when blended.

So a semi transparent red pixel will for example add 50% red, which is correct.
But a almost transparent pixel (say at 5%), will still add 50% of color to the blend... which is wrong

If anyone knows where the blending is done, I'll have a look if you want.

Cheers

Ludvig Larsson aka Valmond
ps. to fix it, I 'painted' the almost transparent pixels with the background color (you won't see a difference in Photoshop, but in the gamy you will), this won't fix the problem if the background changes so it's kind of useless often (or use 50% transparent pixels and not less, which is cumbersome).
